lib/logstash/inputs/redis.rb

```diff
@@ -73,38 +94,79 @@ def register
       @data_type = 'list'
     end
 
-    if not @key or not @data_type
+    if !@key || !@data_type
       raise RuntimeError.new(
         "Must define queue, or key and data_type parameters"
       )
     end
     # end TODO
 
-    @logger.info("Registering Redis", :identity => identity)
+    @redis_builder ||= method(:internal_redis_builder)
+
+    # just switch on data_type once
+    if @data_type == 'list' || @data_type == 'dummy'
+      @run_method = method(:list_runner)
+      @teardown_method = method(:list_teardown)
+    elsif @data_type == 'channel'
+      @run_method = method(:channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    elsif @data_type == 'pattern_channel'
+      @run_method = method(:pattern_channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    end
+
+    # TODO(sissel, boertje): set @identity directly when @name config option is removed.
+    @identity = @name != 'default' ? @name : "#{@redis_url} #{@data_type}:#{@key}"
+    @logger.info("Registering Redis", :identity => @identity)
   end # def register
 
-  # A string used to identify a Redis instance in log messages
-  # TODO(sissel): Use instance variables for this once the @name config
-  # option is removed.
-  private
-  def identity
-    @name || "#{@redis_url} #{@data_type}:#{@key}"
+  def run(output_queue)
+    @run_method.call(output_queue)
+  rescue LogStash::ShutdownSignal
+    # ignore and quit
+  end # def run
+
+  def teardown
+    @shutdown_requested = true
+    @teardown_method.call
   end
 
+  # private methods -----------------------------
   private
-  def connect
-    redis = Redis.new(
+
+  def batched?
+    @batch_count > 1
+  end
+
+  # private
```

Comment:
True.  This is simply a convenience to illustrate the change to public method sections and private ones.  Should be remove before PR is merged.